numpy/core/src/multiarray/descriptor.c

```diff
@@ -1672,17 +1648,17 @@ _convert_from_bytes(PyObject *obj, PyArray_Descr **at)
         goto fail;
     }
 
+    PyArray_Descr *ret;
     if ((check_num == NPY_NOTYPE + 10) ||
-            (*at = PyArray_DescrFromType(check_num)) == NULL) {
+            (ret = PyArray_DescrFromType(check_num)) == NULL) {
         PyErr_Clear();
         /* Now check to see if the object is registered in typeDict */
         if (typeDict == NULL) {
             goto fail;
         }
         PyObject *item = NULL;
 #if defined(NPY_PY3K)
-        PyObject *tmp;
-        tmp = PyUnicode_FromEncodedObject(obj, "ascii", "strict");
+        PyObject *tmp = PyUnicode_FromEncodedObject(obj, "ascii", "strict");
         if (tmp == NULL) {
             goto fail;
```

Comment:
```suggestion
            return NULL;
```
This seems like it was an error, unless setting a new error when one already is set is always OK (I thought it was not on some python versions at least). So maybe we should just clear the error. But maybe its also fine as is.

Reply:
> setting a new error when one already is set is always OK

I'm pretty sure we do it all over the place when replacing an exception with a different one. I'd rather not add that to the pile of bugs I need to think about in this patch.

> So maybe we should just clear the error.

I thought about this too, but decided I'd rather leave things the way they currently are for these cleanup patches, and only start fixing them if I get around to improving the error messages.

The change you suggest would result in a `UnicodeDecodeError` instead of a `TypeError`, which probably isn't helpful to users.

Reply:
Yeah, the DeprecationWarning was what tripped me first. Just keep it as is, for a moment I was just wondering if you changed it earlier accidentally, but doubt that.

Reply:
Nope, this was here from the beginning, and was of the things I was deliberately steering clear of.